File: Python/underautomation/universal_robots/socket_communication/internal/socket_communication_server_base.py
import typing
from underautomation.universal_robots.socket_communication.socket_client import SocketClient
from underautomation.universal_robots.internal.ur_service_base import URServiceBase
import clr
import os
clr.AddReference(os.path.realpath(os.path.join(os.path.dirname(__file__), "..", "..",  'lib', 'UnderAutomation.UniversalRobots.dll')))
from UnderAutomation.UniversalRobots.SocketCommunication.Internal import SocketCommunicationServerBase as socket_communication_server_base

class SocketCommunicationServerBase(URServiceBase):
	def __init__(self, _internal = 0):
		if(_internal == 0):
			self._instance = socket_communication_server_base()
		else:
			self._instance = _internal
	def socket_client_connection(self, handler):
		class Wrapper :
			def __init__(self, _internal):
				self._instance = _internal
		self._instance.SocketClientConnection+= lambda sender, request : handler(Wrapper(sender), Wrapper(request))
	def socket_get_var(self, handler):
		class Wrapper :
			def __init__(self, _internal):
				self._instance = _internal
		self._instance.SocketGetVar+= lambda sender, request : handler(Wrapper(sender), Wrapper(request))
	def socket_request(self, handler):
		class Wrapper :
			def __init__(self, _internal):
				self._instance = _internal
		self._instance.SocketRequest+= lambda sender, request : handler(Wrapper(sender), Wrapper(request))
	def socket_client_disconnection(self, handler):
		class Wrapper :
			def __init__(self, _internal):
				self._instance = _internal
		self._instance.SocketClientDisconnection+= lambda sender, request : handler(Wrapper(sender), Wrapper(request))
	def start(self, port: int) -> None:
		self._instance.Start(port)
	def stop(self) -> None:
		self._instance.Stop()
	def socket_write(self, message: str) -> None:
		self._instance.SocketWrite(message)
	@property
	def connected_clients(self) -> typing.List[SocketClient]:
		return [SocketClient(x) for x in self._instance.ConnectedClients]
	@property
	def enabled(self) -> bool:
		return self._instance.Enabled
	@property
	def port(self) -> int:
		return self._instance.Port

File: SocketCommunication/Internal/SocketCommunicationServerBase.cs
//              WARNING
// This file is an empty shell containing only public C# items.
// The internal code is hidden; to access it, you need to obtain a Source licence of the library.

using Internal;
using SocketCommunication;

namespace SocketCommunication.Internal {
	/// <summary>
	/// Base for Socket communication server
	/// </summary>
	public class SocketCommunicationServerBase : URServiceBase, ISocketHandler {

		/// <summary>
		/// Event handler of a robot that connects with socket_open()
		/// </summary>
		public delegate void SocketClientConnectionEventHandler(object sender, SocketClientConnectionEventArgs request);

		/// <summary>
		/// Event handler when the robot socket disconnects
		/// </summary>
		public delegate void SocketClientDisconnectionEventHandler(object sender, SocketClientDisconnectionEventArgs request);

		/// <summary>
		/// Event handler of a socket message sent with socket_get_var()
		/// </summary>
		public delegate void SocketGetVarEventHandler(object sender, SocketGetVarEventArgs request);

		/// <summary>
		/// Event handler of a socket message received from robot
		/// </summary>
		public delegate void SocketRequestEventHandler(object sender, SocketRequestEventArgs request);

		/// <summary>
		/// List of all connected clients. One robot can open multiple sockets.
		/// </summary>
		public SocketClient[] ConnectedClients { get; }

		/// <summary>
		/// Starts socket server. Robot can connect with URScript function socket_open()
		/// </summary>
		/// <param name="port">Socket server port</param>
		public void Start(int port)
		{
			// Source is hidden, a Source licence is needed to access internal code...
		}

		/// <summary>
		/// Is the socket server enabled
		/// </summary>
		public bool Enabled { get; }

		/// <summary>
		/// Socket server local port
		/// </summary>
		public int Port { get; }

		/// <summary>
		/// Event raised when a robot connects with URScript function socket_open()
		/// </summary>
		public event SocketCommunicationServerBase.SocketClientConnectionEventHandler SocketClientConnection;

		/// <summary>
		/// Event raised when the robot calls socket_get_var()
		/// </summary>
		public event SocketCommunicationServerBase.SocketGetVarEventHandler SocketGetVar;

		/// <summary>
		/// Event raised when a message is received from robot
		/// </summary>
		public event SocketCommunicationServerBase.SocketRequestEventHandler SocketRequest;

		/// <summary>
		/// Event raised when the robot socket disconnects
		/// </summary>
		public event SocketCommunicationServerBase.SocketClientDisconnectionEventHandler SocketClientDisconnection;

		/// <summary>
		/// Disable local socket server and disconnect all connected clients
		/// </summary>
		public void Stop()
		{
			// Source is hidden, a Source licence is needed to access internal code...
		}

		/// <summary>
		/// Write a socket message to the robot. The robot should be connected with socket_open()
		/// </summary>
		/// <param name="message">Message to send to the robot</param>
		public void SocketWrite(string message)
		{
			// Source is hidden, a Source licence is needed to access internal code...
		}
	}
}
